### DISCUSSION TITLE
Extension/plugins for Signature validation

### DISCUSSION THREAD

**Discussion by nunogoncalocastro-vodafone** [emqx/emqx#14009] [Q&A] [answered]
Hi guys, recently we have been looking to see how we could implement a *custom signature validation* on incoming messages; but we are not sure if we understood correctly how the Plugins / Extensions / ExHooks properly work, and we are somewhat confused on this topic; specifically because some of the content we read was 4y old and might be outdated.

- Our main assumption was that, by default, plugins are implemented in *Erlang*, but other programming languages (e.g. Python, Java, ...) could be used to create **ExHooks**, if the [**Extension Hook** plugin](https://github.com/emqx/emqx-exhook) was installed?
  - Another assumption is that as a pre-requisite **Python3 runtime** should be installed in the same server where EMQX would be running? If that's the case, and considering that we are deploying EMQX existing images (e.g. EMQX:5) to an EKS Cluster, and they don't contains python, we would need to create a new image based on the existing ones (with the added runtimes)?
![image](https://github.com/user-attachments/assets/df40339a-829a-46b6-9419-2895b412b912)
- It is also our assumption that after installing the Extension Hook plugin, we could then create our own python ExHooks and just include them in a folder configured in the drivers.path setting (see table below), correct?
![image](https://github.com/user-attachments/assets/630db8a2-e9bd-40d8-9796-2066165348f2)
  - How (or where) do we define/include these configurations? in which file/location? Can they be provided as part of an Helm Chart? if so, how?

- How does this **EMQX ExHook** plugin relates with the [**EMQX Extension Python SDK**](https://github.com/emqx/emqx-extension-python-sdk)? Are they completely different plugins, not related in anyway but overlap with the same purpose? or does one depend on the other and they are interlinked?
  - What's the difference between this **Python SDK** and the previous one discussed above?
  - On the documentation, it is indicated the need to install the python library *emqx-extension-sdk*. Is this a requirement on the EMQX "server" as well? So we would need python runtime (with pip) and this python library for running this plugin?
    ```
    pip3 install emqx-extension-sdk
    ```
  - Also, on the documentation, there's a reference to loading the plugins through emqx_ctl, but no "load" option is available
    ```  
    emqx_ctl plugins load emqx_extension_hook
    ```
![image](https://github.com/user-attachments/assets/56f4c0e6-f42d-408b-8f42-aac26a64e5cf)

I know there are a lot of questions, but any help would be appreciated.
Many thanks.

**Comment by id** [comment, marked as answer]
ExHook is built-in v5 and is called [ExProto Gateway](https://docs.emqx.com/en/emqx/latest/gateway/exproto.html).
Your application can be in whatever language, just need to expose gRPC endpoint for EMQX.

> Another assumption is that as a pre-requisite Python3 runtime should be installed in the same server where EMQX would be running

No.

> It is also our assumption that after installing the Extension Hook plugin, we could then create our own python ExHooks and just include them in a folder configured in the drivers.path setting (see table below), correct?

No, for using ExProto, you don't need to install anything in EMQX, you just need to run your application somewhere.

> How does this EMQX ExHook plugin relates with the [EMQX Extension Python SDK](https://github.com/emqx/emqx-extension-python-sdk)?

EMQX Extension Python SDK is not supported anymore. Plugins extending EMQX functionality can be written only in Erlang https://docs.emqx.com/en/emqx/latest/extensions/plugins.html#develop-emqx-plugins.

**Reply by nunogoncalocastro-vodafone** [reply]
Hi @id quick follow-up question on this topic.
Is there any future plans (or roadmap) for supporting plugins written in other languages besides Erlang?

**Reply by id** [reply]
Nothing on the roadmap so far.

**Reply by nunogoncalocastro-vodafone** [reply]
Thanks for replying back quickly.
